### PR TITLE
Allow to parse configuration file for third-party programs and start the programs

### PR DESCRIPTION
**Pull request by kakurasan**
For part of #103.

This PR allows users to start *already installed* TrucksBook Client (`TB Client.exe`) and other third-party programs.
* *No automated setup for TrucksBook in this PR*
* ~~If `C:\Program Files (x86)\TrucksBook Client\TB Client.exe` is present and readable, `truckersmp-cli` automatically start it before the game~~
* ~~If `--without-trucksbook` is given, `truckersmp-cli` doesn't start it~~
* ~~According to https://github.com/truckersmp-cli/truckersmp-cli/pull/159#issuecomment-841293257, TrucksBook needs sleep -> The default is 60 seconds (can be changed by `--sleep-before-start` option)~~
* Users of third-party programs need to write valid configuration (see the example below)
* Users can specify alternative configuration file (the default is `$XDG_CONFIG_HOME/truckersmp-cli/truckersmp-cli.conf`) by using `-c`/`--configfile` option

An example for TrucksBook users:
```
# Section name must start with "thirdparty."
[thirdparty.trucksbook]
# Path to TB Client.exe (Required)
executable = ~/.local/share/truckersmp-cli/Euro Truck Simulator 2/prefix/pfx/dosdevices/c:/Program Files (x86)/TrucksBook Client/TB Client.exe
# Waiting time before starting game (Optional, default:0)
# Actual waiting time is the maximum value of all thirdparty.* sections
wait = 45
```

### CODE REVIEW

**Comment by Lucki** [comment]
Now this is getting a bit specific for something we've decided not to support because of missing functionality.

Same as in before I thought of this being more generally and just expose our internal `--executable` to the outside together with an optional sleep.
Advantages would be that TrucksBook don't have to be installed in the prefix (it can run without "installation" by simple unpacking) and would allow to also run other similar stuff that might be out there.

**Comment by kakurasan** [comment]
> Same as in before I thought of this being more generally and just expose our internal `--executable` to the outside together with an optional sleep.

If we choose the "general" way, TrucksBook users would need to specify the path *every time*.

Now I'm considering using a directory for third-party programs: `truckersmp-cli` starts the third-party programs which are placed in `$XDG_DATA_HOME/truckersmp-cli/thirdparty/[program_name]/` when the user specifies `--with-thirdparty-programs=` option.

```
[truckersmp-cli]-+-[Euro Truck Simulator 2]
                 +-[American Truck Simulator]
                 +-[TruckersMP]
                 +-[Proton]
                 +-[dlls]
                 +-[wine-discord-ipc-bridge]
                 +-[thirdparty]-+-[program1]-+-program1.exe
                                |            +- ...
                                |
                                +-[program2]-+-program2.exe
                                             +- ...
```

Usage (Draft): `--with-thirdparty-programs={none,all,directory_name...}`
* `--with-thirdparty-programs=none`: Don't start third-party programs (Default)
* `--with-thirdparty-programs=all`: Start all the programs (`*.exe`) placed in `$XDG_DATA_HOME/truckersmp-cli/thirdparty/[program_name]/`
* `--with-thirdparty-programs=program1,program2`: Start the program placed in `$XDG_DATA_HOME/truckersmp-cli/thirdparty/program1/` and `$XDG_DATA_HOME/truckersmp-cli/thirdparty/program2/`

Example for TrucksBook:
```
[truckersmp-cli]-+-[Euro Truck Simulator 2]
                 +-[American Truck Simulator]
                 +-[TruckersMP]
                 +-[Proton]
                 +-[dlls]
                 +-[wine-discord-ipc-bridge]
                 +-[thirdparty]-+-[trucksbook]-+-TB Client.exe
                                | (created     +- ...
                                |     manually)
```
* `--with-thirdparty-programs=trucksbook`: Start `TB Client.exe` with sleeping 60 seconds
* `--with-thirdparty-programs=trucksbook --sleep-before-start=45`: Start `TB Client.exe` with sleeping 45 seconds
* `--with-thirdparty-programs=all`: Start all the programs including `TB Client.exe` with sleeping 60 seconds
* `--with-thirdparty-programs=all --sleep-before-start=45`: Start all the programs including `TB Client.exe` with sleeping 45 seconds

**Comment by Lucki** [comment]
> users would need to specify the path _every time_.

I'd assume users would create a `.desktop` file or start from a launcher like Steam, Lutris or GameHub which would only be once.

Using a subfolder is also a good idea although I would go with simple configuration files similar to a `…/truckersmp-cli.d/program{1,2,…}.conf`.
The advantages I see with config files are:
* The program itself can (almost) be anywhere
* We don't have to guess which is the correct executable in a folder (uninstall.exe)
* We wouldn't even need a new parameter, just always scan the folder (maybe for disabling a parameter could be useful)
* Waiting time can be configured on a per program base and is extendable with other options

-------
~~~
[truckersmp-cli]-+-[Euro Truck Simulator 2]
                 +-[American Truck Simulator]
                 +-[TruckersMP]
                 +-[Proton]
                 +-[dlls]
                 +-[wine-discord-ipc-bridge]
                 +-[thirdparty/whatever]-+-program1.conf
                                         +-program2.conf
                                         +-program3.conf
                                         +- ...
~~~

Simple configuration file example in yaml (be it whatever config language):
~~~ yaml
exe: /path/to/program.exe
sleep: 30
~~~

**Comment by kakurasan** [comment]
We can use [configparser](https://docs.python.org/3/library/configparser.html) module and multiple third-party configurations can be stored in a single configuration file (e.g. `$XDG_CONFIG_HOME/truckersmp-cli/truckersmp-cli.conf`).

Configuration file sample:
```ini
[mainsection]
foo = yes
bar = no

[thirdparty.program1]
executable = /path/to/program1.exe
wait = 30

[thirdparty.program2]
executable = /path/to/program2.exe
wait = 5

[thirdparty.program3]
# No "executable"
wait = 1

[thirdparty.program4]
executable = /path/to/program4.exe
# No "wait"

[thirdparty.program5]
executable = /path/to/program5.exe
# Invalid "wait"
wait = invalid_value
```
Parser sample:
```py
#! /usr/bin/env python3

import configparser
import sys


def main():
    if len(sys.argv) < 2:
        sys.exit("Usage: {} [CONFIGFILE]".format(__file__))
    cfg = configparser.ConfigParser()
    cfg.read(sys.argv[1])
    total_wait = 0
    thirdparty_sects = []
    for sect in cfg.sections():
        if sect.startswith("thirdparty."):
            thirdparty_sects.append(sect)
    for sect in thirdparty_sects:
        try:
            print("executable={}".format(cfg[sect]["executable"]), end=", ")
        except KeyError:
            print("Ignoring section '{}' (No 'executable')".format(sect))
            continue
        try:
            total_wait += int(cfg[sect]["wait"])
            print("wait={}".format(cfg[sect]["wait"]))
        except (KeyError, ValueError):
            print("wait=(Invalid or missing)")
    print("Total wait time: {}".format(total_wait))


if __name__ == "__main__":
    main()
```
Output:
```
executable=/path/to/program1.exe, wait=30
executable=/path/to/program2.exe, wait=5
Ignoring section 'thirdparty.program3' (No 'executable')
executable=/path/to/program4.exe, wait=(Invalid or missing)
executable=/path/to/program5.exe, wait=(Invalid or missing)
Total wait time: 35
```

I guess that the users of third-party programs don't always use them. If we choose to use configuration files, we should allow users to specify another configuration file that doesn't contain `thirdparty.*` sections (e.g. Add `-c`/`--config-file=` option).

File: truckersmp_cli/configfile.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        self._thirdparty_executables = []
+        parser = configparser.ConfigParser()
+        parser.read(configfile)
+        for sect in parser.sections():
+            # get data from valid thirdparty.* sections
+            if not sect.startswith("thirdparty.") or "executable" not in parser[sect]:
+                continue
+            try:
+                wait = int(parser[sect]["wait"])
+                self._thirdparty_total_wait += wait
```
Comment (Lucki) [inline truckersmp_cli/configfile.py:30]: Doesn't using the max value here makes more sense?

If progA needs 20s wait and progB 30s wait and we're starting all at once we've finished the waiting time for progA already in waiting for progB.

**Review by Lucki — APPROVED** [review]
Should be good. Guess we'll have to revisit the readme soon.

File: truckersmp_cli/configfile.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+            # get data from valid thirdparty.* sections
+            if not sect.startswith("thirdparty.") or "executable" not in parser[sect]:
+                continue
+            try:
+                wait = int(parser[sect]["wait"])
+            except (KeyError, ValueError):
+                wait = 0  # invalid or missing
+            self._thirdparty_wait = max(wait, self._thirdparty_wait)
+            self._thirdparty_executables.append(
+                os.path.expanduser(parser[sect]["executable"]))
```
Comment (Lucki) [inline truckersmp_cli/configfile.py:34]: I'm not so sure about this one as this only expands `~` and leaves the other variables untouched (`$XDG…`, `$HOME`).
There's `expandvars()` which handles these but leaves `~` untouched. So we would have to use both together but that's still isn't resolving unset variables.

I guess we should make absolute path mandatory here.

Comment (kakurasan) [reply truckersmp_cli/configfile.py]: > I guess we should make absolute paths mandatory here.

Do we need to check whether the given path is absolute path?

Even if `os.path.isabs(given_path)` returns `False`, Windows style absolute paths (e.g. `C:\Program Files (x86)\TrucksBook Client\TB Client.exe`) work.

```
# When the program is located in the game prefix
executable = C:\Program Files (x86)\TrucksBook Client\TB Client.exe
```

Regarding relative paths: What about treating them as paths relative to our data directory (`Dir.truckersmp_cli_data`, `$XDG_DATA_HOME/truckersmp-cli`)?

```
[thirdparty.program1]
# $XDG_DATA_HOME/truckersmp-cli/program1/program1.exe
executable = program1/program1.exe
```

* `os.path.isabs(given_path)` returns `True` -> Use the given path
* `os.path.isabs(given_path)` returns `False`:
    * The given path is Windows style absolute path -> Use the given path
    * Not Windows style absolute path -> `os.path.join(Dir.truckersmp_cli_data, given_path)`

Comment (Lucki) [reply truckersmp_cli/configfile.py]: I wasn't even aware windows paths work here.

> * `os.path.isabs(given_path)` returns `True` -> Use the given path
> * `os.path.isabs(given_path)` returns `False`:
>   * The given path is Windows style absolute path -> Use the given path
>   * Not Windows style absolute path -> `os.path.join(Dir.truckersmp_cli_data, given_path)`

This seems like a good compromise.

**Comment by kakurasan** [comment]
> Guess we'll have to revisit the readme soon.

@Lucki Are you interested in writing the new section (in another PR)?

**Comment by Lucki** [comment]
Yes

**Comment by Lucki** [comment]
While looking at #208 it seems you want them all in one configuration file while I thought more of a set of switchable configuration files. This is also valid but this may be lacking differentiation between the game modes now. E.g. `thirdparty.ets2mp.prog1` and `thirdparty.ets2.prog1`.

File: truckersmp_cli/configfile.py [hunk 3]
```diff
@@ -28,6 +29,34 @@ def __init__(self, configfile):
...
+                if Args.ets2:  # ets2mp
+                    if (sect.startswith("thirdparty.ets2.")
+                            or sect.startswith("thirdparty.ats.")
+                            or sect.startswith("thirdparty.atsmp.")):
+                        continue
+                else:          # atsmp
+                    if (sect.startswith("thirdparty.ets2mp.")
+                            or sect.startswith("thirdparty.ets2.")
+                            or sect.startswith("thirdparty.ats.")):
+                        continue
```
Comment (Lucki) [inline truckersmp_cli/configfile.py]: This section looks overly complicated.

What about something like
```suggestion
            # get data from valid thirdparty.* sections
            # only use configurations for the specified game
            #   [thirdparty.all.prog1]    -> all games
            #   [thirdparty.ets2mp.prog1] -> ETS2MP
            #   [thirdparty.ets2.prog1]   -> ETS2
            #   [thirdparty.atsmp.prog1]  -> ATSMP
            #   [thirdparty.ats.prog1]    -> ATS
            if not sect.startswith("thirdparty.all") or not sect.startswith("thirdparty." + Args.game) or "executable" not in parser[sect]:
                continue
```

**Review by Lucki — CHANGES_REQUESTED** [review]

**Review by Lucki — APPROVED** [review]

File: truckersmp_cli/configfile.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+                    # DOS/Windows style absolute path: use the given path
+                    self._thirdparty_executables.append(exe_path)
+                else:
+                    # relative path: assume it's relative to our data directory
+                    self._thirdparty_executables.append(
+                        os.path.join(Dir.truckersmp_cli_data, exe_path))
+
+    @property
+    def thirdparty_executables(self):
+        """Rerurn third-party programs paths."""
```
Comment (Lucki) [inline truckersmp_cli/configfile.py:60]:
```suggestion
        """Return third-party program paths."""
```

File: truckersmp_cli/configfile.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        return self._thirdparty_executables
+
+    @property
+    def thirdparty_wait(self):
+        """Rerurn waiting time for third-party programs."""
```
Comment (Lucki) [inline truckersmp_cli/configfile.py:65]:
```suggestion
        """Return waiting time for third-party programs."""
```

**Comment by kakurasan** [comment]
> Rerurn -> Return

Thanks!

**Pull request merged**